Sources/SwiftLexicalLookup/ScopeImplementations.swift

```diff
@@ -0,0 +1,289 @@
+      }
+    case .memberBlock:
+      return statements.flatMap { codeBlockItem in
+        LookupName.getNames(from: codeBlockItem.item)
+      }
+    }
+  }
+
+  /// Returns names matching lookup using provided file
+  /// scope handling configuration (by default: `memberBlockUpToLastDecl`).
+  ///
+  /// Example usage:
+  /// ```swift
+  /// class a {}
+  /// class b {
+  ///   // <--
+  /// }
+  /// let c = 0
+  /// class d {}
+  /// if true {}
+  /// class e {}
+  /// let f = 0
+  /// ```
+  /// According to different configurations,
+  /// names available at the marked place are:
+  /// - for `fileScopeNameIntroductionStrategy` - a, b, c, d
+  /// - for `memberBlock` - a, b, c, d, e, f
+  /// - for `codeBlock` - a
+  public func lookup(
+    for name: String?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig
+  ) -> [LookupResult] {
+    let names = introducedNames(using: config.fileScopeHandling)
+      .filter { introducedName in
+        introducedName.isAccessible(at: syntax) && (name == nil || introducedName.refersTo(name!))
+      }
+
+    return [.fromFileScope(self, withNames: names)]
+  }
+}
+
+@_spi(Experimental) extension CodeBlockSyntax: ScopeSyntax {
+  /// Names introduced in the code block scope
+  /// accessible after their declaration.
+  public var introducedNames: [LookupName] {
+    statements.flatMap { codeBlockItem in
+      LookupName.getNames(from: codeBlockItem.item, accessibleAfter: codeBlockItem.endPosition)
+    }
+  }
+}
+
+@_spi(Experimental) extension ForStmtSyntax: ScopeSyntax {
+  /// Names introduced in the `for` body.
+  public var introducedNames: [LookupName] {
+    LookupName.getNames(from: pattern)
+  }
+}
+
+@_spi(Experimental) extension ClosureExprSyntax: ScopeSyntax {
+  /// All names introduced by the closure signature.
+  /// Could be closure captures or (shorthand) parameters.
+  ///
+  /// Example:
+  /// ```swift
+  /// let x = { [weak self, a] b, _ in
+  ///   // <--
+  /// }
+  /// ```
+  /// During lookup, names available at the marked place are:
+  /// `self`, a, b.
+  public var introducedNames: [LookupName] {
+    let captureNames =
+      signature?.capture?.children(viewMode: .sourceAccurate).flatMap { child in
+        if let captureList = child.as(ClosureCaptureListSyntax.self) {
+          captureList.children(viewMode: .sourceAccurate).flatMap { capture in
+            LookupName.getNames(from: capture)
+          }
+        } else {
+          LookupName.getNames(from: child)
+        }
+      } ?? []
+
+    let parameterNames =
+      signature?.parameterClause?.children(viewMode: .sourceAccurate).flatMap { parameter in
+        if let parameterList = parameter.as(ClosureParameterListSyntax.self) {
+          parameterList.children(viewMode: .sourceAccurate).flatMap { parameter in
+            LookupName.getNames(from: parameter)
+          }
+        } else {
+          LookupName.getNames(from: parameter)
+        }
```

Comment:
This is using the implicit return from `if` expressions, right? swift-syntax needs to be able to compile with Swift 5.7 so we can’t use `if` expressions yet.

```suggestion
        if let parameterList = parameter.as(ClosureParameterListSyntax.self) {
          return parameterList.children(viewMode: .sourceAccurate).flatMap { parameter in
            LookupName.getNames(from: parameter)
          }
        } else {
          return LookupName.getNames(from: parameter)
        }
```